src/flexalloc_xnvme_env.c

```diff
@@ -192,61 +192,59 @@ fla_xne_sync_seq_w_naddrs(struct xnvme_dev * dev, const uint64_t slba, const uin
 // Read or write out a larger chunk striped across multiple objects
 int
 fla_xne_sync_strp_seq_x(struct xnvme_dev *dev, const uint64_t offset, uint64_t nbytes,
-                        void const *buf, struct fla_sync_strp_params *sp, bool write)
+                        void const *buf, struct fla_sync_strp_params const * const sp, bool write)
 {
+  int err;
   uint32_t nsid = xnvme_dev_get_nsid(dev);
   struct xnvme_queue *queue = NULL;
-  int err;
   struct fla_async_cb_args cb_args = { 0 };
+  struct xnvme_cmd_ctx * ctx;
   const struct xnvme_geo *geo = xnvme_dev_get_geo(dev);
-  uint32_t strp_num = sp->strp_num;
-  uint32_t strp_sz = sp->strp_sz;
-  uint64_t obj_len = sp->obj_len;
-  uint32_t strp_blks = (strp_sz / geo->lba_nbytes) - 1;
-  uint64_t bytes_xfer = 0, strp_offset = offset;
+  uint32_t strp_blks = (sp->strp_nbytes / geo->lba_nbytes) - 1;
+  uint64_t bytes_to_xfer = nbytes, strp_offset = offset, strp_to_xfer;
   char *strp_buf = (char *)buf;
 
-  // Nbytes must be a multiple of strp_num * strp_sz
-  if (nbytes % (strp_num * strp_sz))
-  {
-    err = -1;
-    FLA_ERR(err, "Striped write must be a multiple of strp_sz and nbytes");
-    goto exit;
-  }
-
-  // Make sure offset aligns on strp boundary
-  if (offset % strp_sz)
+  if (write)
   {
-    err = -1;
-    FLA_ERR(err, "Striped write offset must be aligned to strp sz");
-    goto exit;
+    if (FLA_ERR(nbytes % (sp->strp_nobjs * sp->strp_nbytes) > 0,
+                "Striped write must be a multiple of sp->strp_nbytes and nbytes")
+        || FLA_ERR(nbytes/sp->strp_nobjs < sp->strp_nbytes,
+                   "Num bytes not large enough for sp->strp_nbytes * sp->strp_nobjs")
+        || FLA_ERR(offset % sp->strp_nbytes > 0,
+                   "Striped write offset must be aligned to strp sz"))
+    {
+      err = -1;
+      goto exit;
+    }
   }
 
-  if (nbytes/strp_num < strp_sz)
+  if (FLA_ERR(nbytes % sp->strp_nbytes > 0,
+              "Striped transfer must be a multiple of stripe number of bytes"))
   {
     err = -1;
-    FLA_ERR(err, "Num bytes not large enough for strp_sz * strp_num");
     goto exit;
   }
 
-  err = xnvme_queue_init(dev, strp_num, 0, &queue);
+  err = xnvme_queue_init(dev, sp->strp_nobjs, 0, &queue);
   if (FLA_ERR(err, "xnvme_queue_init"))
     goto exit;
 
   xnvme_queue_set_cb(queue, fla_async_cb, &cb_args);
 
-  while (bytes_xfer != nbytes)
+  while (bytes_to_xfer > 0)
   {
-
-    for (uint32_t strp=0; strp < strp_num;)
+    strp_to_xfer = bytes_to_xfer / sp->strp_nbytes;
```

Comment:
Do you think we should limit the QD? What if a really large request comes in it is probably worthwhile to limit this to some reasonable value, 128. Let's discuss when we meet.

Reply:
We can use xnvmes capacity to cap the qd and just give it all it can take. I have to double check with simon, but I think this is the right way to do it.

Additinally, have you hit the cap? I think the way we step the transfer and we only use strp_nobj number of ctx in queue, we will never hit the capacity.